lib/driver.js

```diff
@@ -410,460 +391,399 @@ DatastaxDriver.prototype.executeCqlOnDriver = function executeCqlOnDriver(pool,
-        params: params,
-        options: execOptions
-      };
-      stream.emit('error', err);
-    })
-    .pipe(stream);
-};
-
-// TODO: we need to have some sort of incremental retry logic but we will leave
-// this up to the consumer in this case as its a very special cased type thing.
-// We can potentially make it generic in some way later on
-DatastaxDriver.prototype.execCqlStream = function execCqlStream(cql, dataParams, options, stream) {
-  /* istanbul ignore next: no benefit of testing coalesce for keyspace retrieval */
-  var self        = this,
-    keyspace    = (options || {}).keyspace || this.keyspace,
-    consistency = options.consistency || self.poolConfig.consistencyLevel || self.consistencyLevel.one;
-
-  performConnect.call(self, keyspace, false, function (err, pool) {
-    if (err) {
-      return stream.emit(err);
-    }
-
-    if (pool.isReady === false) {
-      pool.waiters.push(function (e) {
-        if (e) {
-          return stream.emit(e);
-        }
-        self.execCqlStream(cql, dataParams, options, stream);
-      });
-      return;
-    }
-
-    self.streamCqlOnDriver(pool, cql, dataParams, consistency, options, stream);
-  });
-};
-
-DatastaxDriver.prototype.execCql = function execCql(cql, dataParams, options, queryRequestId, callback) {
-  /* istanbul ignore next: no benefit of testing coalesce for keyspace retrieval */
-  var self        = this,
-    keyspace    = (options || {}).keyspace || this.keyspace,
-    consistency = options.consistency || self.poolConfig.consistencyLevel || self.consistencyLevel.one;
-
-  performConnect.call(self, keyspace, false, function (err, pool) {
-    if (err) {
-      return void callback(err);
-    }
-
-    if (pool.isReady === false) {
-      pool.waiters.push(function (e) {
-        if (e) {
-          return void callback(e);
-        }
-        self.execCql(cql, dataParams, options, queryRequestId, callback);
-      });
-      return;
+      throw err;
     }
+  }
 
-    // client removes used data entries, so we need to create a copy of the data in case it's needed for retry...
-    var dataCopy       = dataParams.slice(),
-      queryRequestId = uuid.v4();
-    self.emit('queryStarted', queryRequestId, cql, dataCopy, consistency, options);
-    self.executeCqlOnDriver(pool, cql, dataCopy, consistency, options, function (err, results) {
-
-      function execCqlCallback(e, res) {
-        if (Array.isArray(res)) {
-          res = self.getNormalizedResults(res, options);
-        }
-        if (e) {
-          e.cql = cql;
-        }
-        callback(e, res);
-      }
+  async *_execIterable(cql, dataParams, options) {
+    const iterable = await this._execWithRetries(cql, dataParams, options, async (pool, consistency) => {
+      // client removes used data entries, so we need to create a copy of the data in case it's needed for retry...
+      const innerIterable = this._iterateCqlOnDriver(pool, cql, dataParams.slice(), consistency, options);
```

Comment:
This statement may not be true anymore, so may be worth it to test removing the `slice()` here to see if it's still necessary.

Reply:
@scommisso: it turns out that `priam` was not only mutating the array, it was also mutating the param objects as well. Updated code to stop doing these mutations to params.